Documentation/designs/status-subresource.md

```diff
@@ -0,0 +1,436 @@
+    status: "True"
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  shardStatuses:
+  - shardID: "0"
+    replicas: 2
+    updatedReplicas: 2
+    availableReplicas: 2
+    unavailableReplicas: 0
+  - shardID: "1"
+    replicas: 2
+    updatedReplicas: 2
+    availableReplicas: 2
+    unavailableReplicas: 0
+```
+
+Example of a Prometheus resource's status for which some pods are missing due to scheduling issues:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+metadata:
+  name: monitoring
+spec:
+  paused: false
+  replicas: 2
+  shards: 2
+status:
+  paused: false
+  replicas: 4
+  updatedReplicas: 4
+  availableReplicas: 2
+  unavailableReplicas: 2
+  conditions:
+  - type: Available
+    status: "True"
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  - type: Degraded
+    status: "True"
+    reason: "PodsNotReady"
+    message: |-
+      Some pods are not ready:
+      prometheus-monitoring-shard-0-1: 0/6 nodes are available: 2 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate, 2 node(s) had volume node affinity conflict, 2 node(s) were unschedulable.
+      prometheus-monitoring-shard-1-1: 0/6 nodes are available: 2 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate, 2 node(s) had volume node affinity conflict, 2 node(s) were unschedulable.
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  - type: Reconciled
+    status: "True"
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  shardStatuses:
+  - shardID: "0"
+    replicas: 2
+    updatedReplicas: 2
+    availableReplicas: 1
+    unavailableReplicas: 1
+  - shardID: "1"
+    replicas: 2
+    updatedReplicas: 2
+    availableReplicas: 1
+    unavailableReplicas: 1
+```
+
+### Alertmanager
+
+The subresource status for the Alertmanager custom resource definition should
+be very similar to the API defined for the Prometheus CRD. The main difference
+is that we don't need the `shardStatuses` field.
+
+### Thanos Ruler
+
+The subresource status for the Thanos Ruler custom resource definition is
+identical to the Alertmanager CRD.
+
+### Service Monitor
+
+```golang
+type ServiceMonitor struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Specification of desired Service selection for target discovery by
+	// Prometheus.
+	Spec ServiceMonitorSpec `json:"spec"`
+	// 
+	// Most recent observed status of the service monitor.
+	// Populated by the Prometheus operator.
+	// Read-only.
+	Status ServiceMonitorStatus `json:"status"`
+}
+
+type ServiceMonitorStatus struct {
+	// The current state of the service monitor.
+	// +patchMergeKey=type
+	// +patchMergeStrategy=merge
+	// +optional
+	Conditions []ServiceMonitorStatusCondition `json:"conditions,omitempty"`
+
+	// The list of resources that the service monitor is bound to.
+	// +patchMergeKey=name
+	// +patchMergeStrategy=merge
+	// +optional
+	Bindings []ServiceMonitorBindings `json:"bindings,omitempty"`
```

Comment:
Does it make sense to have `ServiceMonitorBindings` in the `Prometheus` status? As a user, I would expect to mostly troubleshoot by inspecting `Prometheus` resources.

Reply:
My concern was that the Prometheus status might become too big if/when you have many service/pod monitors. But I agree that having to go to each service monitor might be less user-friendly.

Having said that, it makes sense to reflect the number of bound objects in the Prometheus status field (there's a TODO at the beginning of this section):

```
status:
  rejectedServiceMonitors: 10
  selectedServiceMonitors: 120
```

Reply:
Yes, this also came to my mind. Ok let's start with bindings in the `ServiceMonitor`. It's easier to add fields than to take them away.